src/compiler/checker.ts

```diff
@@ -10387,16 +10387,30 @@ module ts {
             }
         }
 
-        function getFirstNonAmbientClassOrFunctionDeclaration(symbol: Symbol): Declaration {
+        function getFirstClassOrFunctionDeclaration(symbol: Symbol, nonAmbientOnly: boolean): Declaration {
             let declarations = symbol.declarations;
             for (let declaration of declarations) {
-                if ((declaration.kind === SyntaxKind.ClassDeclaration || (declaration.kind === SyntaxKind.FunctionDeclaration && nodeIsPresent((<FunctionLikeDeclaration>declaration).body))) && !isInAmbientContext(declaration)) {
+                if ((declaration.kind === SyntaxKind.ClassDeclaration || (declaration.kind === SyntaxKind.FunctionDeclaration && nodeIsPresent((<FunctionLikeDeclaration>declaration).body))) && !(nonAmbientOnly && isInAmbientContext(declaration))) {
                     return declaration;
                 }
             }
             return undefined;
         }
 
+        function inSameLexicalScope(node1: Node, node2: Node) {

```

Comment:
This should be correct, as the class and module have differing block scopes. I do not want to elide the `var` when I emit the module in your example, otherwise the emit would be:
```js
var container;
(function (container) {
  class clodule { }
  container.clodule = clodule;
})(container || (container = {});

var container;
(function (container) {
  (function (clodule) {
    ;
  })(clodule = container.clodule || (container.clodule = {})); // error: 'clodule' is not defined
})(container || container = {});
```

We need to emit the `var` here, to ensure a local variable for `clodule` is in scope:

```js
var container;
(function (container) {
  class clodule { }
  container.clodule = clodule;
})(container || (container = {});

var container;
(function (container) {
  var clodule; // will not collide with the lexical declaration above.
  (function (clodule) {
    ;
  })(clodule = container.clodule || (container.clodule = {})); 
})(container || container = {});
```